### PR TITLE
[parity-util-mem] add support for mimalloc

### LINKED ISSUE #180: [parity-util-mem] add support for mimalloc
https://github.com/microsoft/mimalloc seems to be a better choice than jemalloc (time- and memory-wise). We would probably need a low-level rust crate wrapping its C API first, like jemalloc-sys.

### PR DESCRIPTION
**Pull request by ordian**
Fixes #180

### CODE REVIEW

File: parity-util-mem/Cargo.toml [hunk 1]
```diff
@@ -11,11 +11,12 @@ cfg-if = "0.1.6"
...
 dlmalloc = { version = "0.1", features = ["global"], optional = true }
 wee_alloc = { version = "0.4", optional = true }
-mimallocator = { version = "0.1", optional = true }
+mimallocator = { version = "0.1", features = ["secure", "stats"], optional = true }
+mimalloc-sys = { version = "0.1", features = ["secure", "stats", "override"], optional = true }
 
 elastic-array = { version = "0", optional = true }
 ethereum-types = { version = "0", optional = true }
-parking_lot = { version = "0", optional = true }
+parking_lot = { version = "0.7", optional = true }
```
Comment (dvdplm) [inline parity-util-mem/Cargo.toml]: I think @cheme had a reason for keeping the version very open ended here. You sure you need to nail it down like this?

Comment (ordian) [reply parity-util-mem/Cargo.toml]: This is a temp workaround for cargo not picking the right version (probably a bug) of parking_lot and failing to compile because of that.

Comment (cheme) [reply parity-util-mem/Cargo.toml]: Yes, I must admit that I had to set it manually in parity-eth cargo.lock, I do not think the change will break in substrate (already a lot of parking lot different version there), but yes the intent was that parking lot implementation can use multiple version (this is not really a future proof mechanism : breaking change on parking lot would mean having to manage multiple version explicitely which may prove being challenging).

**Comment by ordian** [comment]
The build is failing on macOS, but I'm not sure if it's because of outdated mac toolchain on travis or actually legit.

**Comment by dvdplm** [comment]
Test passes here (osx) and prints a bunch of stats, is that intended?

**Comment by ordian** [comment]
> prints a bunch of stats, is that intended?

removed "stats" feature of mimalloc, should be gone now

**Comment by dvdplm** [comment]
There's a compilation error on CI: missing headers?

**Comment by dvdplm** [comment]
cc @TriplEight is this a CI issue you reckon?

**Comment by ordian** [comment]
now macOS tests pass, but linux fail :man_facepalming:

**Comment by TriplEight** [comment]
@dvdplm no, doesn't look like a CI issue.

**Comment by ordian** [comment]
@dvdplm the CI is green now

File: parity-util-mem/src/allocators.rs [hunk 2]
```diff
@@ -95,6 +99,13 @@ cfg_if! {
 			jemallocator::usable_size(ptr)
 		}
 
+	} else if #[cfg(feature = "mimalloc-global")] {
+
+		/// Use of mimalloc usable size C function through mimallocator crate call.
+		pub unsafe extern "C" fn malloc_usable_size(ptr: *const c_void) -> usize {
+			mimalloc_sys::mi_usable_size(ptr as *mut _)
```
Comment (ordian) [inline parity-util-mem/src/allocators.rs]: mimalloc doesn't actually mutate the value `ptr` points to, but requires a mut pointer in the API

Comment (dvdplm) [reply parity-util-mem/src/allocators.rs]: That sounds like something we should put in a comment in the code.

File: .travis.yml [hunk 3]
```diff
@@ -5,12 +5,16 @@ branches:
 matrix:
   include:
     - os: linux
+      dist: xenial
```
Comment (dvdplm) [inline .travis.yml]: Why was this needed? Maybe a comment for posterity?

File: parity-util-mem/Cargo.toml [hunk 4]
```diff
@@ -11,6 +11,8 @@ cfg-if = "0.1.6"
 malloc_size_of_derive = "0.1.0"
 dlmalloc = { version = "0.1", features = ["global"], optional = true }
 wee_alloc = { version = "0.4", optional = true }
+mimallocator = { version = "0.1", features = ["secure"], optional = true }
```
Comment (dvdplm) [inline parity-util-mem/Cargo.toml]: What is `"secure"`?
